pkg/kubelet/network/bridge/bridge.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	if _, err = plugin.cniConfig.AddNetwork(plugin.netConfig, rt); err != nil {
+		return fmt.Errorf("Error adding container to network: %v", err)
+	}
+
+	return nil
+}
+
+func (plugin *bridgeNetworkPlugin) TearDownPod(namespace string, name string, id kubetypes.DockerID) error {
+	runtime, ok := plugin.host.GetRuntime().(*dockertools.DockerManager)
+	if !ok {
+		return fmt.Errorf("Bridge execution called on non-docker runtime")
+	}
+	netnsPath, err := runtime.GetNetNs(id.ContainerID())
+	if err != nil {
+		return err
+	}
+
+	rt := buildCNIRuntimeConf(name, namespace, id.ContainerID(), netnsPath)
+	if err != nil {
+		return fmt.Errorf("Error building CNI config: %v", err)
+	}
+
+	if err := plugin.cniConfig.DelNetwork(plugin.netConfig, rt); err != nil {
+		return fmt.Errorf("Error removing container from network: %v", err)
+	}
+
+	return nil
+}
+
+// TODO: Use the addToNetwork function to obtain the IP of the Pod. That will assume idempotent ADD call to the plugin.
+// Also fix the runtime's call to Status function to be done only in the case that the IP is lost, no need to do periodic calls
+func (plugin *bridgeNetworkPlugin) Status(namespace string, name string, id kubetypes.DockerID) (*network.PodNetworkStatus, error) {
+	runtime, ok := plugin.host.GetRuntime().(*dockertools.DockerManager)
+	if !ok {
+		return nil, fmt.Errorf("Bridge execution called on non-docker runtime")
+	}
+	ipStr, err := runtime.GetContainerIP(string(id), DefaultInterfaceName)
+	if err != nil {
+		return nil, err
+	}
+	ip, _, err := net.ParseCIDR(strings.Trim(ipStr, "\n"))
+	if err != nil {
+		return nil, err
+	}
+	return &network.PodNetworkStatus{IP: ip}, nil
+}
+
+func buildCNIRuntimeConf(podName string, podNs string, podInfraContainerID kubecontainer.ContainerID, podNetnsPath string) *libcni.RuntimeConf {
+	glog.V(4).Infof("Bridge: using netns path %v", podNetnsPath)
+	glog.V(4).Infof("Bridge: using podns path %v", podNs)
+
+	return &libcni.RuntimeConf{
+		ContainerID: podInfraContainerID.ID,
+		NetNS:       podNetnsPath,
+		IfName:      DefaultInterfaceName,
+	}
+}
+
+func (plugin *bridgeNetworkPlugin) PodCIDRChanged(cidr string) {
+	var err error
+
+	if err := ensureIPTablesMasqRule(); err != nil {
+		err = fmt.Errorf("Error on adding ip table rules: %v", err)
+		glog.Error(err)
+	}
+
+	if len(cidr) == 0 {
+		err = fmt.Errorf("ConfigureCBR0 requested, but PodCIDR not set. Will not configure CBR0 right now")
+		glog.Warning(err)
+	} else if err := plugin.updateCbr0Cidr(cidr); err != nil {
+		err = fmt.Errorf("Error configuring cbr0: %v", err)
+		glog.Error(err)
+	}
+
+	plugin.host.SetNetworkState(err)
+}
+
+func (plugin *bridgeNetworkPlugin) updateCbr0Cidr(podCIDR string) error {
+	glog.V(5).Infof("PodCIDR is set to %q", podCIDR)
+	_, cidr, err := net.ParseCIDR(podCIDR)
+	if err != nil {
+		return err
+	}
+	// Set cbr0 interface address to first address in IPNet
+	cidr.IP.To4()[3] += 1
+	if err := ensureCbr0CIDR(cidr); err != nil {
```

Comment:
hmm, please find a way to remove this, calling it defeats the purpose of this experiment. ensure will always save us. I want to use the stock bridge so we don't have to maintain this code.